pkg/agent/clean/clean.go

```diff
@@ -0,0 +1,325 @@
+/*
+The cleanup is designed to get a cluster that was imported to Rancher disconnected
+and cleanup some objects
+
+Remove the cattle-system namespace which houses the agent used to talk
+to Rancher
+
+Remove Rancher labels, finalizers and annotations from all namespaces
+
+If the cluster was imported to Rancher 2.1 or later cleanup roles, rolebindings,
+clusterRoles and clusterRoleBindings
+*/
+
+package clean
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/rancher/rancher/pkg/controllers/user/nslabels"
+	"github.com/sirupsen/logrus"
+	apierror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var (
+	// List of namespace labels that will be removed
+	nsLabels = []string{
+		nslabels.ProjectIDFieldLabel,
+	}
+
+	// List of namespace annotations that will be removed
+	nsAnnotations = []string{
+		"cattle.io/status",
+		"field.cattle.io/creatorId",
+		"lifecycle.cattle.io/create.namespace-auth",
+		nslabels.ProjectIDFieldLabel,
+	}
+)
+
+func Cluster() error {
+	// The sleep allows Rancher server time to finish updating ownerReferences
+	logrus.Info("Starting sleep for 1 min....")
+	time.Sleep(time.Duration(1) * time.Minute)
+	logrus.Info("Starting cluster cleanup")
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	var errors []error
+
+	err = removeCattleNamespace(client)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	nsErr := cleanupNamespaces(client)
+	if len(nsErr) > 0 {
+		errors = append(errors, nsErr...)
+	}
+
+	crbErr := cleanupClusterRoleBindings(client)
+	if len(crbErr) > 0 {
+		errors = append(errors, crbErr...)
+	}
+
+	cbErr := cleanupRoleBindings(client)
+	if len(cbErr) > 0 {
+		errors = append(errors, cbErr...)
+	}
+
+	crErr := cleanupClusterRoles(client)
+	if len(crErr) > 0 {
+		errors = append(errors, crErr...)
+	}
+
+	rErr := cleanupRoles(client)
+	if len(rErr) > 0 {
+		errors = append(errors, rErr...)
+	}
+
+	if len(errors) > 0 {
+		return processErrors(errors)
+	}
+
+	return deleteJob(client)
+}
+
+func removeCattleNamespace(client *kubernetes.Clientset) error {
+	logrus.Info("Attempting to remove cattle-system namespace")
+	return tryUpdate(func() error {
+		ns, err := client.CoreV1().Namespaces().Get("cattle-system", metav1.GetOptions{})
+		if err != nil {
+			if apierror.IsNotFound(err) {
+				return nil
+			}
+			return err
+		}
+		if len(ns.Finalizers) > 0 {
+			ns.Finalizers = []string{}
+		}
+
+		ns, err = client.CoreV1().Namespaces().Update(ns)
```

Comment:
Putting the comment here as an example, but it applies to you entire PR:
can you add more logging? Log every time you make an update or or delete statement